### PR TITLE
Apply FAI §9.1.3 cylinder tolerance + boot-time IGC reprocess

### LINKED ISSUE #42: Cylinder crossing detection ignores FAI Sporting Code tolerance — pilots missing TPs by <20 m at 4 km cylinders
## Summary

The pipeline detects cylinder crossings using exact `radius_m` (`segmentIntersectsCircle` in `src/pipeline.ts`). FAI Sporting Code S7F §9.1.3 requires a **tolerance** to be applied to observation-zone boundaries:

> Tolerance is applied separately to the straight portion and to the semi-circle.

Standard practice in XC scoring (mirroring official FAI tools): **0.5% of the radius, minimum 5 m**. For our usual radii:

| Radius | Tolerance | Effective inner radius for "tag" |
|---|---|---|
| 200 m  | 5 m     | 195 m |
| 500 m  | 5 m     | 495 m |
| 1000 m | 5 m     | 995 m |
| 4000 m | 20 m    | 3980 m |

Without this tolerance we are **stricter than FAI**, which is illegal scoring for any task that calls itself FAI-compliant.

## Concrete prod hit

Ian Heidenberger's submission to "Fly, you fools (Apr/May)" (NWPC 2026): closest approach to GRN-6K (4 km cylinder) was **4001 m** — he missed by **1 m**. With the standard 20 m tolerance he would have correctly tagged TP1 and his last_turnpoint_index / distance_flown_km / score would all jump up. Today he sits at rank 7 with 1.272 km / 300.8 pts; his distance to TP1 alone is enough to bump him into the rank-4-to-6 cluster (~440-460 pts).

## Fix

In `src/shared/task-engine.ts` (or wherever `segmentIntersectsCircle` is consumed), apply the tolerance to the radius used for the *inward* crossing test:

```ts
function effectiveTagRadius(radiusM: number): number {
  const tolerance = Math.max(5, radiusM * 0.005);
  return radiusM + tolerance;  // tag if within radius + tolerance
}
```

Notes:
- Only applies to the "tag" detection (whether a fix segment hits the boundary). The *optimised route distance* still uses the strict radius — pilots get credit for tagging within tolerance, but the route geometry is unchanged.
- SSS direction-of-crossing logic stays the same; tolerance just widens the threshold radius by 5 m / 0.5% on the inward check.
- For GOAL_LINE goals (the "D-shape" semi-circle + line) the tolerance applies separately to the chord and the semi-circle per the spec — current code uses one path for both, may need splitting.

## Backfill

Once the fix lands, all existing scored attempts need re-processing (the new logic could re-tag previously-missed cylinder crossings). Add a `RESCORE_ALL_TASKS` boot-time job (or a one-shot script) that:

1. Re-runs the pipeline against each `flight_submissions.igc_data` blob.
2. Replaces the corresponding `flight_attempts` rows.
3. Calls `rebuildTaskResults` for each touched task.

This is similar to the `REPROCESS_ALL_SUBMISSIONS` handler that was registered but never enqueued (and was removed in PR #31). Would re-introduce here as a one-shot startup task, then remove again after deploy completes.

## Test plan

- Unit test in `src/pipeline.test.ts`: a fix segment whose closest approach to the cylinder centre is `radius - 4 m`, `radius + 1 m`, `radius + 6 m`, `radius + 25 m` for a 4 km cylinder. Expected tag results: yes, yes, yes, no.
- Same for a 500 m cylinder with 1 m / 4 m / 6 m miss distances. Expected: yes, yes, no.
- Regression test that re-processing Ian's IGC against this task produces `last_turnpoint_index >= 1`.

## References

- FAI Sporting Code S7F (XC Scoring 2025 v1.0): https://www.fai.org/sites/default/files/civl/documents/sporting_code_s7_f_-_xc_scoring_2025_v1.0.pdf §9.1.3
- Project notes: `CLAUDE.md` "Goal line observation zone" section

### PR DESCRIPTION
**Pull request by james-willis**
Closes #42.

## Summary

The cylinder crossing detection in `detectAttempts` used strict `radius_m` for the in/out check. FAI Sporting Code S7F §9.1.3 prescribes a tolerance of **`max(5 m, 0.5% × radius)`** — pilots get the benefit of the doubt for tags. A 4 km cylinder gets a 20 m tolerance.

**Concrete prod hit fixed by this PR:** Ian Heidenberger's submission to *Fly, you fools (Apr/May)* had a closest approach of **4001 m** to GRN-6K (0.8 m miss). With the new tolerance he tags TP1 cleanly and his `distance_flown_km` / `last_turnpoint_index` / score all jump up.

## What changes

- New `tagToleranceM(r) = max(5, r * 0.005)` helper applied to:
  - The SSS exit gate (in/out check + intersection radius)
  - In-cylinder check on every other turnpoint
  - `segmentIntersectsCircle` radius for graze-through detection
  - The goal-line chord and goal-arc semi-circle (per §9.1.3 separately)
  - HAF ground-state classifier's "fix inside cylinder" test
- The optimised-route geometry (`optimiseRoute`, `computePartialDistanceKm`) continues to use the strict radius — tolerance only governs *tags*.
- New `SCORER_VERSION = '1.1'` exported from `pipeline.ts`. Upload writes it into `flight_attempts.scorer_version` (column pre-existed, was unused).
- New `src/reprocess.ts`: `reprocessStaleSubmissions(db)` finds every flight_submission with attempts at an older scorer_version, re-parses the stored `igc_data` blob, replaces `flight_attempts` + `turnpoint_crossings` in a per-submission txn, then rebuilds `task_results` once per touched task. Per-submission errors are logged and skipped.
- Wired into `server.ts` boot, before the existing `rebuildTaskResults` loop. No-op when every attempt is current; ~50 ms per submission when stale.

## After deploy

Every existing IGC gets reprocessed exactly once. Pilots may see score deltas as previously-missed cylinder tags are now credited. Expected biggest mover: Ian (currently rank 7 with 1.272 km / 300.8 pts on Fly, you fools — should jump into the rank 4-6 cluster around 440-460 pts).

## Test plan

- [x] Unit tests cover the tolerance contract (5 m floor below 1000 m radius, 0.5% above).
- [x] `npm test` (194/194 pass).
- [x] `npm run typecheck`.
- [ ] After deploy: verify boot logs show `[reprocess] re-running pipeline against N submissions for SCORER_VERSION=1.1` and `[reprocess] done`.
- [ ] After deploy: confirm Ian's `last_turnpoint_index >= 1` for sub `8e5f6a22-...` on `Fly, you fools` (TP1 = GRN-6K).
- [ ] After deploy: watch the standings on `https://league.nwparagliding.com/leagues/nwpc` for new score deltas.

## Notes

- Companion to #43 (cylinder *rendering* uses a uniform-radius SVG `<circle>` — drawn ~2.7 m oversized at 47.5° lat). Filing as separate PR; tolerance is the actual scoring fix and ships first.

### CODE REVIEW

File: src/reprocess.ts [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+  db.transaction(() => {
+    // Drop old attempt + crossing rows for this submission. flight_attempts
+    // has soft-delete but the upload path overwrites in place; do the same
+    // here. turnpoint_crossings has no soft-delete column so hard-delete.
+    db.prepare(
+      `DELETE FROM turnpoint_crossings
+       WHERE attempt_id IN (SELECT id FROM flight_attempts WHERE submission_id = ?)`,
+    ).run(sub.id);
+    db.prepare(`DELETE FROM flight_attempts WHERE submission_id = ?`).run(sub.id);
```
[the remark this answers is not in the document]
Comment (james-willis) [reply src/reprocess.ts:206]: Fixed in 8587abb: the per-submission transaction now does `DELETE FROM task_results WHERE task_id` first, then deletes the old `turnpoint_crossings` and `flight_attempts`, then inserts new attempts/crossings, and finally calls `rebuildTaskResults` inside the same txn to repopulate `task_results`. This avoids the FK violation while keeping each submission atomic.

File: src/reprocess.ts [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    }
+  }
+
+  for (const taskId of affectedTasks) {
+    try {
+      rebuildTaskResults(db, taskId);
+    } catch (err) {
+      console.error(`[reprocess] rebuildTaskResults failed for task ${taskId}:`, err);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (james-willis) [reply src/reprocess.ts]: Good catch — fixed in 8587abb. Dropped the post-loop `affectedTasks` rebuild entirely; `rebuildTaskResults` now runs inside each per-submission transaction so reprocess is responsible for its own task_results, and the boot-time loop in server.ts isn't doing redundant work.

File: src/pipeline.ts [hunk 3]
```diff
@@ -369,21 +400,28 @@ export function detectAttempts(fixes: Fix[], task: TaskDefinition): Result<Attem
...
+        // semi-circle; we widen both by the same tolerance value.
+        const tChord = segmentIntersectsGoalLine(aLocal, bLocal, { x: 0, y: 0 }, effectiveR, bearingDeg);
+        const tArc = segmentEntersGoalSemiCircle(aLocal, bLocal, { x: 0, y: 0 }, effectiveR, bearingDeg);
         if (tChord !== null || tArc !== null) {
           crossT = Math.min(tChord !== null ? tChord : Infinity, tArc !== null ? tArc : Infinity);
           crossed = true;
         }
-      } else if (distA < tp.radiusM) {
-        // Already inside this cylinder — count as immediately achieved
+      } else if (distA < effectiveR) {
```
[the remark this answers is not in the document]
Comment (james-willis) [reply src/pipeline.ts]: Fixed in 8587abb: changed `distA < effectiveR` to `distA <= effectiveR` so the code matches the §9.1.3 contract documented in the comment (a fix at distance d is "inside" iff d ≤ r + tolerance).

File: src/reprocess.ts [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    db.prepare(`DELETE FROM task_results WHERE task_id = ?`).run(sub.task_id);
+
+    // Drop old attempt + crossing rows for this submission. turnpoint_crossings
+    // has no soft-delete column so hard-delete; flight_attempts has one but
+    // the upload path overwrites in place — do the same here.
+    db.prepare(
+      `DELETE FROM turnpoint_crossings
+       WHERE attempt_id IN (SELECT id FROM flight_attempts WHERE submission_id = ?)`,
+    ).run(sub.id);
+    db.prepare(`DELETE FROM flight_attempts WHERE submission_id = ?`).run(sub.id);
```
[the remark this answers is not in the document]
Comment (james-willis) [reply src/reprocess.ts:206]: Good catch — addressed in 717b859. Added a pre-flight check that throws `submission has turnpoint_overrides — manual reconciliation required` if any `turnpoint_overrides` rows reference this submission's attempts. The per-submission try/catch in `reprocessStaleSubmissions` then logs and skips that submission, leaving its existing `flight_attempts` and the audit-record overrides intact. Migrating overrides to new crossing IDs is a future concern once the override endpoint actually lands; for now the audit row stays authoritative.

File: src/reprocess.ts [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+      nowIso,
+      sub.id,
+    );
+
+    // Re-create task_results for this task from the new flight_attempts,
+    // inside the same txn so the FK constraint stays satisfied at commit.
+    // better-sqlite3 transactions are reentrant via savepoints, so the
+    // nested rebuildTaskResults transaction is safe here.
+    rebuildTaskResults(db, sub.task_id);
+  })();
```
[the remark this answers is not in the document]
Comment (james-willis) [reply src/reprocess.ts]: Agreed and adopted in 717b859. Switched to a targeted `DELETE FROM task_results WHERE best_attempt_id IN (SELECT id FROM flight_attempts WHERE submission_id = ?)` (just enough to satisfy the FK before the attempt delete), and dropped the per-submission `rebuildTaskResults` call entirely — `server.ts` already sweeps every non-deleted task with `rebuildTaskResults` immediately after `reprocessStaleSubmissions` returns, which re-derives the full table once instead of O(N²).

File: src/reprocess.ts [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+    lat: tp.latitude,
+    lng: tp.longitude,
+    radiusM: tp.radius_m,
+    type: tp.type as TurnpointDef['type'],
+    forceGround: tp.force_ground === 1,
+    goalLineBearingDeg: tp.goal_line_bearing_deg ?? undefined,
+  }));
+
+  // Goal times from *other* submissions on the same task — needed for time-points
+  // scoring, even though rebuildTaskResults will overwrite the score columns.
```
[the remark this answers is not in the document]
Comment (james-willis) [reply src/reprocess.ts]: Fixed in 717b859. Updated the comment to reflect what's actually happening: existing goal times feed the pipeline's per-attempt time-points calculation during this submission's run, and the boot-time `task_results` sweep in `server.ts` is the single point that re-derives per-task scores afterward. The per-attempt point columns are not touched by `rebuildTaskResults` — that's an artefact slated for removal in #32.

File: src/pipeline.ts [hunk 7]
```diff
@@ -369,21 +400,29 @@ export function detectAttempts(fixes: Fix[], task: TaskDefinition): Result<Attem
       let crossed = false;
       let crossT: number | null = null;
 
+      // Apply FAI §9.1.3 tolerance to the inward tag detection. The boundary
+      // is treated as r + tolerance for the in/out check; the optimised route
+      // and partial-distance geometry continue to use the strict radius.
+      const effectiveR = tp.radiusM + tagToleranceM(tp.radiusM);
+
```
[the remark this answers is not in the document]
Comment (james-willis) [reply src/pipeline.ts:408]: Added in 0137dd2 — 11 regression tests in `tests/goal-line.test.ts`:\n\n- 8 `segmentNearGoalLine` tests covering the new perpendicular-tolerance contract: actual crossing, exact touch, just-short perpendicular, beyond-tolerance perpendicular, parallel within tolerance, parallel beyond tolerance, near-endpoint within tolerance, near-endpoint beyond tolerance.\n- 3 `segmentIntersectsCircle` × `tagToleranceM` tests covering the cylinder boundary: 4 m short of a 400 m cyl tags, 8 m short doesn't, 15 m short of a 4000 m cyl tags via the 0.5 %-of-radius floor.\n\nThis exercises the in-loop `effectiveR = r + tagToleranceM(r)` logic via the same exported geometry helpers that `detectAttempts` calls, which is what was missing.

File: src/reprocess.ts [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+  // Distinct submissions whose flight_attempts are at an older scorer version.
+  // Soft-deleted submissions/attempts are excluded.
+  const stale = db
+    .prepare(
+      `SELECT DISTINCT fs.id, fs.task_id, fs.user_id, fs.league_id, fs.igc_data
+       FROM flight_submissions fs
+       JOIN flight_attempts fa ON fa.submission_id = fs.id
+       WHERE fa.scorer_version != ? AND fs.deleted_at IS NULL AND fa.deleted_at IS NULL`,
+    )
+    .all(SCORER_VERSION) as SubmissionRow[];
```
[the remark this answers is not in the document]
Comment (james-willis) [reply src/reprocess.ts:61]: Declining for now as premature at our scale. Production currently has on the order of ~50 `flight_attempts` rows in the active season; the boot scan is well under a millisecond and the query runs once per process start. A partial index keyed on `!= ?` isn't expressible in SQLite (parameter inside index expr) and a plain index on a column that is overwhelmingly one value isn't very selective. Happy to add it if/when we cross the size where it becomes a real cost — flagging this comment so it's easy to find via the audit trail.

File: src/pipeline.ts [hunk 9]
```diff
@@ -369,21 +400,29 @@ export function detectAttempts(fixes: Fix[], task: TaskDefinition): Result<Attem
...
       if (tp.type === 'GOAL_LINE' && cachedGoalBearing != null) {
         const bearingDeg = cachedGoalBearing;
-        const tChord = segmentIntersectsGoalLine(aLocal, bLocal, { x: 0, y: 0 }, tp.radiusM, bearingDeg);
-        const tArc = segmentEntersGoalSemiCircle(aLocal, bLocal, { x: 0, y: 0 }, tp.radiusM, bearingDeg);
+        // Per §9.1.3 tolerance applies separately to the chord and the
+        // semi-circle; we widen both by the same tolerance value.
+        const tChord = segmentIntersectsGoalLine(aLocal, bLocal, { x: 0, y: 0 }, effectiveR, bearingDeg);
+        const tArc = segmentEntersGoalSemiCircle(aLocal, bLocal, { x: 0, y: 0 }, effectiveR, bearingDeg);
```
[the remark this answers is not in the document]
Comment (james-willis) [reply src/pipeline.ts:416]: Good catch — fixed in 0137dd2. New `segmentNearGoalLine` helper treats the chord as a stadium (chord segment Minkowski-summed with a disk of radius `tagToleranceM(r)`), which gives the chord both perpendicular thickness and rounded ends. The call site now passes strict `r` as the half-length and tolerance separately, so we no longer double-count by stretching `halfLengthM` and the semi-circle radius gets its own `r + tolerance` extension as before. Boundary cases covered by the new tests above (parallel < tolerance tags; parallel > tolerance doesn't; perpendicular short < tolerance tags; perpendicular short > tolerance doesn't).

File: src/pipeline.ts [hunk 10]
```diff
@@ -528,6 +570,80 @@ export function segmentIntersectsGoalLine(
...
+  const rad = (bearingDeg * Math.PI) / 180;
+  const dx = Math.sin(rad) * halfLengthM;
+  const dy = Math.cos(rad) * halfLengthM;
+  const p1: Point2D = { x: lineMidpoint.x - dx, y: lineMidpoint.y - dy };
+  const p2: Point2D = { x: lineMidpoint.x + dx, y: lineMidpoint.y + dy };
+
+  // Min distance between two non-intersecting segments lives at one of the
+  // four endpoint-to-other-segment distances. Pick the smallest one whose
+  // distance is within tolerance, and use its t-on-AB for time interpolation.
+  const candidates: Array<{ dist: number; tOnAB: number }> = [
```
[the remark this answers is not in the document]
Comment (james-willis) [reply src/pipeline.ts]: Good catch — fixed in 2e63e91. Rewrote `segmentNearGoalLine` as an analytic first-entry calculation:\n\n1. Rotate AB into the chord-aligned frame (chord along x-axis from -L to +L).\n2. Test `isInsideStadium(A)` — return 0 if so.\n3. Otherwise collect every `t ∈ [0,1]` at which AB crosses the stadium boundary (top edge `y = +tol`, bottom edge `y = -tol`, both end-cap circles `(x ∓ L)² + y² = tol²`), guarding the cap intersections to only count the part outside the rectangle so an inside-rectangle disc-arc point doesn't masquerade as a later "first entry".\n4. Return `Math.min(...)` of the candidates.\n\nTests updated for the new `t` values (a clean perpendicular crossing now returns t≈0.45, not 0.5 — that's the first moment the track is at `y = -tol` rather than at the chord centreline) and added one new test asserting that AB punching straight through the stadium returns the entry t rather than the exit.

File: src/pipeline.ts [hunk 11]
```diff
@@ -285,7 +310,13 @@ export function detectAttempts(fixes: Fix[], task: TaskDefinition): Result<Attem
   const sss = task.turnpoints[0];
 
   // ── Step 1: Find all SSS outward crossings ──────────────────────────────────
+  // FAI Sporting Code S7F §9.1.3 tolerance: a fix at distance d from the
+  // cylinder centre is "inside" iff d ≤ r + tolerance. The boundary effectively
+  // shifts outward by `tolerance(r)`. For SSS this means the exit gate widens
```
[the remark this answers is not in the document]
Comment (james-willis) [reply src/pipeline.ts]: Fixed in 73c1459 — comment now reads `tagToleranceM(r)` to match the helper used on the very next line.

File: tests/goal-line.test.ts [hunk 12]
```diff
@@ -194,6 +200,110 @@ describe('goal D-shape: chord + semi-circle together', () => {
   });
 });
 
+// ─────────────────────────────────────────────────────────────────────────────
+// segmentNearGoalLine — FAI §9.1.3 perpendicular tolerance on the chord
+// ─────────────────────────────────────────────────────────────────────────────
+
+describe('segmentNearGoalLine', () => {
+  // East-west chord at origin, half-length 100 m, bearing 90°. Tolerance is
+  // tagToleranceM(200) === 5m for the cylinder these endpoints came from.
```
[the remark this answers is not in the document]
Comment (james-willis) [reply tests/goal-line.test.ts]: You're right — fixture uses `R = 100`, so the underlying cylinder is 100 m, not 200 m. Updated the comment in 73c1459 to read `tagToleranceM(100) = 5 m` and explain that 0.5 % of 100 m (= 0.5 m) is below the 5 m floor.

File: src/reprocess.ts [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+    // rebuild here would be O(N²) when many stale submissions share a task;
+    // the boot sweep in server.ts re-derives the whole table once afterward.
+    db.prepare(
+      `DELETE FROM task_results
+       WHERE best_attempt_id IN (SELECT id FROM flight_attempts WHERE submission_id = ?)`,
+    ).run(sub.id);
+
+    // Drop old attempt + crossing rows for this submission. turnpoint_crossings
+    // has no soft-delete column so hard-delete; flight_attempts has one but
+    // the upload path overwrites in place — do the same here.
```
[the remark this answers is not in the document]
Comment (james-willis) [reply src/reprocess.ts]: Right — the upload handler only ever INSERTs, never overwrites. Comment rewritten in 18d8b08 to describe what reprocess is actually doing: replacing stale rows wholesale because we're re-scoring the same IGC under a newer SCORER_VERSION (soft-delete would leave the old + new attempts visible alongside each other).

File: tests/goal-line.test.ts [hunk 14]
```diff
@@ -194,6 +200,111 @@ describe('goal D-shape: chord + semi-circle together', () => {
...
+    expect(t).toBeNull();
+  });
+
+  it('uses the 0.5% floor for large cylinders (4000 m → 20 m tolerance)', () => {
+    const r = 4000;
+    const effectiveR = r + tagToleranceM(r); // 4020
+    // 15 m short of the strict edge — within 20 m tolerance
+    const t = segmentIntersectsCircle({ x: 5000, y: 0 }, { x: 4015, y: 0 }, effectiveR);
+    expect(t).not.toBeNull();
+  });
```
[the remark this answers is not in the document]
Comment (james-willis) [reply tests/goal-line.test.ts]: Good catch — at 4000 m, 0.5 % yields 20 m, which is above the 5 m floor, so the floor isn't applied. Renamed in 18d8b08 to "uses the 0.5% rule above the floor (4000 m → 20 m tolerance)".

File: src/reprocess.ts [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+
+export async function reprocessStaleSubmissions(db: Database): Promise<{ reprocessed: number; failed: number }> {
+  // Distinct submissions whose flight_attempts are at an older scorer version.
+  // Soft-deleted submissions/attempts are excluded.
+  const stale = db
+    .prepare(
+      `SELECT DISTINCT fs.id, fs.task_id, fs.user_id, fs.league_id, fs.igc_data
+       FROM flight_submissions fs
+       JOIN flight_attempts fa ON fa.submission_id = fs.id
+       WHERE fa.scorer_version != ? AND fs.deleted_at IS NULL AND fa.deleted_at IS NULL`,
```
[the remark this answers is not in the document]
Comment (james-willis) [reply src/reprocess.ts]: Good catch — fixed in f47d50c. Switched to an EXISTS subquery so SQLite dedupes on `fs.id` alone and the `igc_data` BLOB is read once per submission instead of being compared as part of the DISTINCT key:\n\n```sql\nSELECT fs.id, fs.task_id, fs.user_id, fs.league_id, fs.igc_data\nFROM flight_submissions fs\nWHERE fs.deleted_at IS NULL\n  AND EXISTS (\n    SELECT 1 FROM flight_attempts fa\n    WHERE fa.submission_id = fs.id\n      AND fa.scorer_version != ?\n      AND fa.deleted_at IS NULL\n  )\n```
